pkg/stream-handler.go

```diff
@@ -62,41 +63,46 @@ func (h *streamHandler) RunStream(ctx context.Context, req *backend.RunStreamReq
+		defer cancelFn()
+
 		msg, err := reader.Next(deadline)
-		cancelFn()
-		if msg != nil {
-			log.DefaultLogger.Info("Received msg.")
-			if err == nil {
-				log.DefaultLogger.Info(fmt.Sprintf("    Message: %s", msg.Payload()))
-				var notification = Notification{}
-				err = json.Unmarshal(msg.Payload(), &notification)
-				if err == nil {
-					// Work for later; Refactor so that the serialization below is happening in the Pulsar message receiver
-					// Go-routine to reduce work needed if more than one client is connected.
-					labelFrame.Fields[0].Set(0, notification.Time)
-					labelFrame.Fields[1].Set(0, notification.Source)
-					labelFrame.Fields[2].Set(0, notification.Key)
-					labelFrame.Fields[3].Set(0, string(notification.Value))
-					labelFrame.Fields[4].Set(0, notification.Message)
-					labelFrame.Fields[5].Set(0, notification.Exception.Message)
-					labelFrame.Fields[6].Set(0, notification.Exception.StackTrace)
-					log.DefaultLogger.Info("Sending notification to " + req.PluginContext.User.Login)
-					err = sender.SendFrame(labelFrame, data.IncludeAll)
-					if err != nil {
-						log.DefaultLogger.Error(fmt.Sprintf("Couldn't send frame: %v", err))
-						return err
-					}
-				} else {
-					log.DefaultLogger.Error(fmt.Sprintf("Could not unmarshall json: %v", err))
-				}
-			} else {
-				log.DefaultLogger.Error(fmt.Sprintf("Couldn't get the message via reader.Next(): %+v", err))
-			}
-			select {
-			case <-ctx.Done():
-				log.DefaultLogger.Info("Grafana sender: DONE")
-				return ctx.Err()
-			}
+		if err != nil {
+			log.DefaultLogger.Error(fmt.Sprintf("Couldn't get the message via reader.Next(): %+v", err))
+			continue
+			// shouldn't we return an error here?
+			// return fmt.Errorf("Couldn't get the message from pulsar: %v", err)
+		}
+
+		if msg == nil {
+			log.DefaultLogger.Error("Received empty msg from pulsar")
+			continue
+			// return fmt.Errorf("got empty msg")
+		}
+
+		log.DefaultLogger.Info(fmt.Sprintf("    Message: %s", msg.Payload()))
+		notification := Notification{}
+		err = json.Unmarshal(msg.Payload(), &notification)
+		if err != nil {
+			log.DefaultLogger.Error(fmt.Sprintf("Could not unmarshal json: %v", err))
+			continue
+		}
+
+		// Work for later; Refactor so that the serialization below is happening in the Pulsar message receiver
+		// Go-routine to reduce work needed if more than one client is connected.
+		labelFrame.Fields[0].Set(0, notification.Time)
+		labelFrame.Fields[1].Set(0, notification.Source)
+		labelFrame.Fields[2].Set(0, notification.Key)
+		labelFrame.Fields[3].Set(0, string(notification.Value))
+		labelFrame.Fields[4].Set(0, notification.Message)
+		labelFrame.Fields[5].Set(0, notification.Exception.Message)
+		labelFrame.Fields[6].Set(0, notification.Exception.StackTrace)
+
+		log.DefaultLogger.Info("Sending notification to " + req.PluginContext.User.Login)
+
+		err = sender.SendFrame(labelFrame, data.IncludeAll)
```

Comment:
Need to be more careful with refactoring. The `select` existed to exit when Grafana detects that the stream is disconnected.

Reply:
good point; will fix that in the evening

Reply:
Refactored a bit, will test it better tomorrow.